### ISSUE TITLE
[Lab 7 Level 4] Expected Output for .printlog()

### ISSUE THREAD

**Issue by vishandi** [nus-cs2030/2021-s2#176]
I don't quite understand the specification needed for the printlog() function of flatMap.

Is the modifications made inside of the input of flatMap (the Function<? super T, ? extends Logger<? extends U>>) also recorded by the Logger? I assume that it is yes, because they are recorded in the test case. But why does the one modification on Logger.make(x) is not recorded (Value unchanged. Value = x)? So what is the expected output of this lines of code: 
`Logger.make(5).flatMap(x -> Logger.make(2 * x).map(y -> y + 2).flatMap(y -> Logger.make(y).map(z -> z * 10))).printlog()`

Will the change on Logger.make(2*x) supposed to be recorded? 

Thank you in advance.

**Comment by High325** [comment]
My interpretation of the implementation of flatMap is that it would always start with ```.flatMap(x -> Logger.make(x).(...))```. This is to say that there would not be a case where ```Logger.make(5).flatMap(x -> Logger.make(2 * x).(...)``` would be used. Personally, I also do not really see any situations where flatMap would be useful in a practical setting but for the implementation of flatMap that they are looking for, flatMap(x -> Logger.make(2 * x)) would not record the change. (at least there's no test case checking for flatMap(x -> Logger.make(2 * x)))

**Comment by eltonngkk** [comment]
I agree with @High325 that ```Logger.make(5).flatMap(x -> Logger.make(2 * x).(...)``` is not very applicable to this implementation, because ideally you only want to initialize the value once at the start, then you log the changes. Thus, the ```(2 * x)``` would be better translated as a map operation. The value of flatMap here is to handle the return type of a stream, you can refer to Recitation 7 to relate better!

**Comment by vishandi** [comment]
Yes, I understand about what flatMap does, what I don't understand is the log keeping. How can we make assumptions like "it would always start with `.flatMap(x -> Logger.make(x).(...))`"? I mean, is it that obvious to make that assumption? Because personally I don't think it is. To put it in another words, is there any justification for the assumption?

**Comment by kathyijaya** [comment]
Hi @vishandi! I think for this lab we can always go with such assumption.
I think we can agree with @High325 and @eltonngkk for this matter. It would not really be intuitive to have a flatMap that initializes a different value as it is something that the method map is supposed to do!
Additionally, output wise, I think it would be different. The method might output something along the line of initializing the value of 2*x when we are actually changing the value of it, which might not be what the Logger is supposed to do.
CMIIW
